cmd/collectors/rest/plugins/shelf/shelf.go

```diff
@@ -0,0 +1,303 @@
+	"time"
+)
+
+type Shelf struct {
+	*plugin.AbstractPlugin
+	data           map[string]*matrix.Matrix
+	instanceKeys   map[string]string
+	instanceLabels map[string]*dict.Dict
+	client         *rest.Client
+	query          string
+}
+
+func New(p *plugin.AbstractPlugin) plugin.Plugin {
+	return &Shelf{AbstractPlugin: p}
+}
+
+func (my *Shelf) Init() error {
+
+	var err error
+	shelfMetric := make(map[string][]string)
+
+	shelfMetric["fans => fan"] = []string{
+		"^^id => fan_id",
+		"^location",
+		"^state => status",
+		"rpm",
+	}
+	shelfMetric["current_sensors => sensor"] = []string{
+		"^^id => sensor_id",
+		"^location",
+		"^state => status",
+		"current => reading",
+	}
+	shelfMetric["frus => psu"] = []string{
+		"^^id => psu_id",
+		"^installed => enabled",
+		//"^location",
+		"^part_number",
+		"^serial_number => serial",
+		"^psu.model => type",
+		"^state => status",
+		"psu.power_drawn => power_drawn",
+		"psu.power_rating => power_rating",
+	}
+	shelfMetric["temperature_sensors => temperature"] = []string{
+		"^^id => sensor_id",
+		"^threshold.high.critical => high_critical",
+		"^threshold.high.warning => high_warning",
+		"^ambient => temp_is_ambient",
+		"^threshold.low.critical => low_critical",
+		"^threshold.low.warning => low_warning",
+		"^state => status",
+		"temperature => reading",
+	}
+	shelfMetric["voltage_sensors => voltage"] = []string{
+		"^^id => sensor_id",
+		"^location",
+		"^state => status",
+		"voltage => reading",
+	}
+
+	if err = my.InitAbc(); err != nil {
+		return err
+	}
+
+	timeout := rest.DefaultTimeout * time.Second
+	if my.client, err = rest.New(conf.ZapiPoller(my.ParentParams), timeout); err != nil {
+		my.Logger.Error().Stack().Err(err).Msg("connecting")
+		return err
+	}
+
+	if err = my.client.Init(5); err != nil {
+		return err
+	}
+
+	my.query = "api/storage/shelves"
+	my.Logger.Info().Msg("plugin connected!")
+
+	my.data = make(map[string]*matrix.Matrix)
+	my.instanceKeys = make(map[string]string)
+	my.instanceLabels = make(map[string]*dict.Dict)
+
+	for attribute, childObj := range shelfMetric {
+
+		objectName := strings.ReplaceAll(attribute, "-", "_")
+
+		if x := strings.Split(attribute, "=>"); len(x) == 2 {
+			attribute = strings.TrimSpace(x[0])
+			objectName = strings.TrimSpace(x[1])
+		}
+
+		my.instanceLabels[attribute] = dict.New()
+
+		my.data[attribute] = matrix.New(my.Parent+".Shelf", "shelf_"+objectName, "shelf_"+objectName)
+		my.data[attribute].SetGlobalLabel("datacenter", my.ParentParams.GetChildContentS("datacenter"))
+
+		exportOptions := node.NewS("export_options")
+		instanceLabels := exportOptions.NewChildS("instance_labels", "")
+		instanceKeys := exportOptions.NewChildS("instance_keys", "")
+		instanceKeys.NewChildS("", "shelf")
+		instanceKeys.NewChildS("", "channel")
+
+		// artificial metric for status of child object of shelf
+		my.data[attribute].NewMetricUint8("status")
+
+		for _, c := range childObj {
+
+			metricName, display := collector.ParseMetricName(c)
```

Comment:
you can move templating method to util and use that here.

Reply:
done, moved to util package.